### PR TITLE
feat: support health check and alert for sites

### LINKED ISSUE #25: [feature] support monitoring and alerting against attack events
See senario at: https://help.aliyun.com/zh/waf/web-application-firewall-3-0/user-guide/configure-waf-alerting?spm=a2c4g.11186623.0.0.23486c1eFMmu08

Can send Email out via Casdoor Email provider

### PR DESCRIPTION
**Pull request by love98ooo**
Fix: https://github.com/casbin/caswaf/issues/25

### CODE REVIEW

File: object/site_cache.go [hunk 1]
```diff
@@ -105,6 +106,9 @@ func refreshSiteMap() error {
 		}
 
 		newSiteMap[strings.ToLower(site.Domain)] = site
+		if site.EnableAlert {
+			healthCheckNeededDomains = append(healthCheckNeededDomains, site.Domain)
```
Comment (nomeguy) [inline object/site_cache.go]: How to delete an item in healthCheckNeededDomains?

File: controllers/account.go [hunk 2]
```diff
@@ -59,3 +59,13 @@ func (c *ApiController) GetAccount() {
 
 	c.ResponseOk(claims, hostname)
 }
+
+func (c *ApiController) GetProviders() {
```
Comment (nomeguy) [inline controllers/account.go]: Not this file

File: object/site_timer.go [hunk 3]
```diff
@@ -77,6 +79,41 @@ func monitorSiteCerts() error {
 	return err
 }
 
+func healthCheck(site *Site, domain string) error {
```
Comment (nomeguy) [inline object/site_timer.go]: Create new Go file to put your code

File: web/src/backend/AccountBackend.js [hunk 4]
```diff
@@ -34,3 +34,10 @@ export function signout() {
     credentials: "include",
   }).then(res => res.json());
 }
+
+export function getProviders() {
```
Comment (nomeguy) [inline web/src/backend/AccountBackend.js]: Create new file

**Review by nomeguy — CHANGES_REQUESTED** [review]

File: object/site_cache.go (hunk 1 above)
Comment (love98ooo) [reply object/site_cache.go:110]: Call refreshSiteMap function

**Review by nomeguy — APPROVED** [review]

File: object/site_timer_health.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	var log string
+	switch site.SslMode {
+	case "HTTPS Only":
+		flag, log = pingUrl("https://" + domain)
+	case "HTTP":
+		flag, log = pingUrl("http://" + domain)
+	case "HTTPS and HTTP":
+		flag, log = pingUrl("https://" + domain)
+		flagHttp, logHttp := pingUrl("http://" + domain)
+		flag = flag || flagHttp
```
Comment (nomeguy) [inline object/site_timer_health.go]: This code is ugly, let ChatGPT write it

File: object/site_timer_health.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		fmt.Println(log)
+		healthCheckTryTimesMap[domain]--
+		if healthCheckTryTimesMap[domain] != 0 {
+			return nil
+		}
+		log = fmt.Sprintf("CasWAF health check failed for domain %s, %s", domain, log)
+		user, err := casdoorsdk.GetUser(site.Owner)
+		if err != nil {
+			fmt.Println(err)
+		}
```
Comment (nomeguy) [inline object/site_timer_health.go:68]: check user == nil

File: object/site_timer_health.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	case "HTTP":
+		flag, log = pingUrl("http://" + domain)
+	case "HTTPS and HTTP":
+		flag, log = pingUrl("https://" + domain)
+		flagHttp, logHttp := pingUrl("http://" + domain)
+		flag = flag || flagHttp
+		log = log + logHttp
+	}
+	if !flag {
+		fmt.Println(log)
```
Comment (nomeguy) [inline object/site_timer_health.go]: Don't print to console anywhere

File: object/site_timer_health.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+var healthCheckTryTimesMap = map[string]int{}
+
+func healthCheck(site *Site, domain string) error {
+	var flag bool
+	var log string
+	switch site.SslMode {
+	case "HTTPS Only":
+		flag, log = pingUrl("https://" + domain)
```
Comment (nomeguy) [inline object/site_timer_health.go]: pingUrl should be only called once or twice

File: object/site_timer_health.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	"time"
+
+	"github.com/casbin/caswaf/util"
+	"github.com/casdoor/casdoor-go-sdk/casdoorsdk"
+)
+
+var healthCheckTryTimesMap = map[string]int{}
+
+func healthCheck(site *Site, domain string) error {
+	var flag bool
```
Comment (nomeguy) [inline object/site_timer_health.go]: Don't name it flag or log

File: object/site_timer_health.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		healthCheckTryTimesMap[domain]--
+		if healthCheckTryTimesMap[domain] != 0 {
+			return nil
+		}
+		log = fmt.Sprintf("CasWAF health check failed for domain %s, %s", domain, log)
+		user, err := casdoorsdk.GetUser(site.Owner)
+		if err != nil {
+			fmt.Println(err)
+		}
+		err = casdoorsdk.SendEmail("CasWAF HealthCheck Check Alert", log, "CasWAF", user.Email)
```
Comment (nomeguy) [inline object/site_timer_health.go]: Support SMS too

File: object/site_timer_health.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func startHealthCheckLoop() {
+	for _, domain := range healthCheckNeededDomains {
+		domain := domain
+		if _, ok := healthCheckTryTimesMap[domain]; ok {
+			continue
+		}
+		healthCheckTryTimesMap[domain] = GetSiteByDomain(domain).AlertTryTimes
+		go func() {
```
Comment (nomeguy) [inline object/site_timer_health.go:86]: Don't use Go routine. Just execute it one by one

File: object/site_timer_health.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+		flag, log = pingUrl("https://" + domain)
+	case "HTTP":
+		flag, log = pingUrl("http://" + domain)
+	case "HTTPS and HTTP":
+		flag, log = pingUrl("https://" + domain)
+		flagHttp, logHttp := pingUrl("http://" + domain)
+		flag = flag || flagHttp
+		log = log + logHttp
+	}
+	if !flag {
```
Comment (nomeguy) [inline object/site_timer_health.go]: Return early

if flag then return/continue

File: object/site_timer_health.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/casbin/caswaf/util"
+	"github.com/casdoor/casdoor-go-sdk/casdoorsdk"
+)
+
+var healthCheckTryTimesMap = map[string]int{}
```
Comment (nomeguy) [inline object/site_timer_health.go:25]: Add this int to Site, store it in DB

File: object/site_timer_health.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+func startHealthCheckLoop() {
+	for _, domain := range healthCheckNeededDomains {
+		domain := domain
+		if _, ok := healthCheckTryTimesMap[domain]; ok {
+			continue
+		}
+		healthCheckTryTimesMap[domain] = GetSiteByDomain(domain).AlertTryTimes
+		go func() {
+			defer func() {
+				if r := recover(); r != nil {
```
Comment (nomeguy) [inline object/site_timer_health.go]: Don't recover

File: object/site_timer_health.go (hunk 10 above)
Comment (nomeguy) [inline object/site_timer_health.go]: You didn't use site.AlertProviders anywhere in Go

File: object/site_timer_health.go [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+					fmt.Printf("[%s] Recovered from healthCheck() panic: %v\n", util.GetCurrentTime(), r)
+				}
+			}()
+			for {
+				site := GetSiteByDomain(domain)
+				if site == nil || !site.EnableAlert {
+					return
+				}
+
+				err := healthCheck(site, domain)
```
Comment (nomeguy) [inline object/site_timer_health.go:94]: You should stop checking a site, when it status is Inactive, or is obvious not configured, like domain and port are empty, etc.

**Review by nomeguy — CHANGES_REQUESTED** [review]

**Review by nomeguy — APPROVED** [review]

**Pull request merged**